### ISSUE TITLE
Mosquitto 5.1.1 is broken.

### ISSUE THREAD

**Issue by adamf663** [home-assistant/addons#1887]
## Environment

- Add-on with the issue: Mosquitto broker
- Add-on release with the issue: 5.1.1
- Last working add-on release (if known): 5.1
- Operating environment (OS/Supervised): hassio

## Problem-relevant configuration

```yaml

```

## Traceback/Error logs

```txt

```

**Comment by LifeBandit666** [comment]
I know OP has left zero information here and it will probably be closed out, but on a surface level I have to agree. Something has gone wrong with this point update. 

I get my whole Home Assistant instance go down overnight. When I try to access it over the network I cant connect. 

After a week of troubleshooting I have narrowed it down to my Mqtt server. What seems to be happening is it loses connection to HA and throws the whole thing off. HA is still running and inaccessible. I have to power cycle to get it up and running again and when I do, I get errors in HA logs about "timed out waiting for mid 2" for Mqtt, and errors in mosquitto about being unable to find HA. 

I also find my Zigbee2mqtt devices are found at startup then immediately lost again (presumably when the time out occurs) and have to restart Zigbee2mqtt a couple of times to get it running again. 

I wish i could prove that the HA hang is due to mosquitto but since it's inaccessible (but still running) I can't. What I can say is that in Recorder I can see the only devices that are unavailable are Zigbee2mqtt and this drops because of mosquitto (there's zero errors in the Zigbee2mqtt logs in debug mode). 

I've restored an old backup of the previous version of Mosquitto to try and combat the issue.

**Comment by LifeBandit666** [comment]
Ok so after rolling back to 5.1 version of Mosquitto I no longer get the timeout errors on startup in the core log and Zigbee2mqtt seems to boot from the get go again (instead of requiring multiple reboots to find Mosquitto) Hopefully this will also fix the issue of my whole HA server becoming unresponsive periodically too. 

This has been a week long hunt to try and find out why my system is dying which I was initially blaming on a new cable for my SSD. I've also tried looking at my Powerline system thinking it was cutting connection to HA.

**Comment by hoopsta1423** [comment]
Yup, I updated last night and the add-on CPU usage spiked up to 150%. All my zwave2mqtt devices stopped working. Rolled back to 5.1 this morning and it's back to normal

**Comment by grantalewis** [comment]
Confirming same. Very strange, sluggish system recently with hundreds of lines similar to 

`2021-02-27 07:18:29 ERROR (MainThread) [homeassistant.components.mqtt] Timed out waiting for mid 98`

in home-assistant.log. After rolling Mosquitto back to 5.1, no more errors in the log, performance back to normal.

**Comment by peter-vanpoucke** [comment]
I also needed to rollback to 5.1. I had the feeling Hassio couldn't log in any longer, but didn't put time investigating. 😔

**Comment by andersmoldin** [comment]
I too seem to have this problem after installing Mosquitto 5.1.1. I did not make a snapshot, so I wonder if it is possible to install a specific version of Mosquitto, or in any other way downgrade to 5.1?

**Comment by willysoft80** [comment]
> Confirming same. Very strange, sluggish system recently with hundreds of lines similar to
> 
> `2021-02-27 07:18:29 ERROR (MainThread) [homeassistant.components.mqtt] Timed out waiting for mid 98`
> 
> in home-assistant.log. After rolling Mosquitto back to 5.1, no more errors in the log, performance back to normal.

I have the same problem and the home assistant has become slow to read the topics

**Comment by realthk** [comment]
> I have the same problem and the home assistant has become slow to read the topics
Yes, that was what I've noticed first: most of my stuff is MQTT-based, and it suddenly took a few seconds to switch on a light through MQTT, controlled by a door- or motion-sensor, also connected through MQTT. I tried restart, then hundreds of "Timed out waiting for mid xxx" messages in the log.

Found this topic, rolled back MQTT to 5.1 and no messages any more, and light is switched on well under a second again.

**Comment by Dinth** [comment]
Yep, same here - none of the things using MQTT in my home works anymore - Openzwave, Ebus, Zigbee2MQTT, Valetudo - all the devices went "Unavailable". I have tried restarting every component but still cant get any of my devices to work. 
Strangle - MQTT addon log shows that addons are connected:
```
1614440933: Client zwave already connected, closing old connection.
1614440933: New bridge connected from 172.30.33.2 as zwave (p2, c1, k60, u'addons').
1614440933: Client blueIris already connected, closing old connection.
1614440933: New client connected from 10.10.1.14 as blueIris (p1, c1, k20, u'xxx').
1614440933: Client mqttjs_c5c88712 already connected, closing old connection.
1614440933: New client connected from 172.30.33.5 as mqttjs_c5c88712 (p2, c1, k60, u'xxx').
1614440933: Client ebusd_21.1_770 already connected, closing old connection.
1614440933: New client connected from 10.10.25.2 as ebusd_21.1_770 (p1, c1, k60, u'xxx').
1614440933: Client blueIris already connected, closing old connection.
1614440933: New client connected from 10.10.1.14 as blueIris (p1, c1, k20, u'xxx').
```
But the devices those components are responsible for are unavailable in HA.
Also when im trying to connect to MQTT using an MQTT client, it doesnt connect anymore:
`1614440958: Socket error on client auto-FBB4A2D6-C2EF-EC2F-C7C9-04379354796B, disconnecting.`

**Comment by JoJa1101** [comment]
Same here, MQTT not working well anymore. Getting a ton of [homeassistant.components.mqtt] Timed out waiting for mid messages

**Comment by davidms12** [comment]
I have one instance running 5.1 and it's working fine. My main machine I upgraded to 5.1.1 - big mistake - all devices through mqtt are now showing as unavailable and tons of "timeouts waiting for mid..."

**Comment by JoJa1101** [comment]
Exactly: looks like this since 5.1.1 

![image](https://user-images.githubusercontent.com/76157868/109420883-1909ce00-79d5-11eb-82ba-9eb59f99c955.png)

**Comment by G3rry71** [comment]
Same issue here! How do I force 5.1.0 installation from Supervisor? Thanks

**Comment by adamf663** [comment]
Don't quote me as I'm guessing.  I had used an old snapshot.
I think a 'docker pull homeassistant/aarch64-addon-mosquitto:5.1' might work.  The only ways I know to access the docker containers is through the portainer addon, or by ssh'ing to -p 22222 root@<your machine>.  I think enabling 22222 also requires an addon.

**Comment by Dinth** [comment]
I think that restoring a snapshot (it is possible to restore JUST the addon) is the only option

**Comment by sir106** [comment]
after restart of home assistant and upgrading i have the same problem.

**Comment by JoJa1101** [comment]
this is srsly annoying right now! I have to switch out lights manually  and close shutters now ^^

**Comment by maxlyth** [comment]
Same symptoms here with 5.1.1 update

In the log I was getting a lot of 'Socket error on client DVES..' errors so I thought this was caused by my recent changes to inter-VLAN routing of Unfi-Wifi config. 

Wasted too many hours on this before I saw this thread and reverted just the Mosquitto part of a recent snapshot which instantly fixed the problem.

**Comment by ahknight** [comment]
Exact same issue.  Everything's broken with 5.1.1.  Luckily had an old snapshot and I'll be trying to restore from that.

**Comment by JoJa1101** [comment]
I did the same. Rolled back only Mosquito to 5.1 and it works qzite well again.

**Comment by alex-savin** [comment]
@JoJa1101 How did you roll back?

**Comment by blacknacoustic** [comment]
> I did the same. Rolled back only Mosquitto to 5.1 and it works well again.

Hey can anyone help me roll back? How would i do that?

**Comment by bsmeding** [comment]
Same here only rollback from old snapshot restored the mqtt function

Would be a nice feature to get a version dropdown by the addons so a rollback would be easier for everyone

**Comment by JoJa1101** [comment]
> @JoJa1101 How did you roll back?

I made a snapshot form my HASS Testsystem (was running on Mosquitto 5.1) and did the partial rollback on my Prod System.

**Comment by alex-savin** [comment]
The issue that socat (which serves as auth point) is being continuously executed again and again!

**Comment by rpitera** [comment]
Glad I found this; I was just about to install and try to migrate from CloudMQTT to a local instance.  I only used CM because I needed an external server to use with an older mobile location method that I don't use anymore since the HA App came out.  Guess I'll subscribe to this thread and wait it out.

**Comment by Mariusthvdb** [comment]
very much an issue here, especially since the mqtt binary_sensors keep triggering the automations on each reload (don't know how to call it but Mqtt gets re-triggered quite often now...)

do we have an add-on dev listening in on this?

**Comment by poudenes** [comment]
Have same issue... It seems that the speed of mqtt is also changed...
I get those errors as well while restart HA.
I see that my zwave2mqtt reaction time is changed with a short delay.

I don't know if this is because of the bug that's how in latest version

**Comment by davidms12** [comment]
Any Devs looking into this? thx.

**Comment by juslex** [comment]
Same issue here

**Comment by davidms12** [comment]
No snapshot here. Is there any way to use command line to downgrade to 5.1? 

ha addons install ???????

thx.

**Comment by blacknacoustic** [comment]
This is my suggestion to anyone who is using Mosquitto inside of there HA instance. I would suggest using or setting up a MQTT broker outside of the HA and linking it inside HA manually using yaml and the ip address.

**Comment by poudenes** [comment]
> This is my suggestion to anyone who is using Mosquitto inside of there HA instance. I would suggest using or setting up a MQTT broker outside of the HA and linking it inside HA manually using yaml and the ip address.
> 
> https://www.home-assistant.io/integrations/mqtt/

3 Days I changed everything from inside to outside. Using a RPi3 for a MQTT Broker (Moquitto) to much power, can move to RPi Zero I guess. And everything is blasting fast again!!!! Even faster then before the brick!

**Comment by adamf663** [comment]
Enough is enough.  Either take 5.1.1 down or fix it!
At the very least provide a back out or workaround method that doesn't require going back to a snapshot.
   

----- Original Message -----

**Comment by nickrout** [comment]
What an angry demanding fellow you turned out to be. First you post an error report without logs or anything useful. Then you post a demanding grumpy post. 

FWIW 5.1.1 is working fine here, amd64 and home assistant supervised.

**Comment by davidms12** [comment]
nickrout you're lucky. It is not working under the arm/Pi installs...

**Comment by ahknight** [comment]
I'm on Proxmox on a Skylake NUC and had the same issue.  It isn't the microarchitecture.

**Comment by Dinth** [comment]
Yep, I'm on R720/ESXi, definitely not related to platform

**Comment by jsb5151** [comment]
Same issue here; 5.1.1 is very unresponsive.  Accepts a connection from MQTT Explorer but takes a loooong while to display the topics & values.   HA Supervised on a x86 compute module.  Never had any issues before 5.1.1.

**Comment by michaelarnauts** [comment]
> Same issue here; 5.1.1 is very unresponsive. Accepts a connection from MQTT Explorer but takes a loooong while to display the topics & values. HA Supervised on a x86 compute module. Never had any issues before 5.1.1.
> 
> Referenced #1897 too - seems like the same issue...

I'm not really sure that #1897 is the same issue. Besides the memory leak, I don't get any timeouts.

One of the changes between 5.1 and 5.1.1 is this: https://github.com/home-assistant/addons/commit/d291f564ced1c19d0cfaa349b9f3b4b91e615151 So it is probably related to that, although I don't know what bashio does.

**Comment by michaelarnauts** [comment]
What do you see when you subscribe to the topic `$SYS/broker/messages/#`? You can do this from the MQTT integration page, or through the cli with mosquitto_sub:

```
mosquitto_sub -v -t '$SYS/broker/messages/#' -u <username> -P <password>`
```

I'm seeing stuff like this:
```
$SYS/broker/messages/stored 4974794
$SYS/broker/messages/received 5020091
$SYS/broker/messages/sent 45801
```

And I suspect that I shouldn't be seeing a high value for `$SYS/broker/messages/stored`. This might explain my memory leak, but maybe it can also explain some of the other issues described here.

**Comment by nickrout** [comment]
I am getting 205 stored against roughly 2,000,000 sent and received.

**Comment by nickrout** [comment]
> although I don't know what bashio does.

> Bashio is a bash function library for use with Home Assistant add-ons. It contains a set of commonly used operations and can be used to be included in add-ons to reduce code duplication across add-ons and therefore making it easier to develop and maintain add-ons.

**Comment by ozbob01** [comment]
I also see the same issues, initially I had both zwave and zigbee on mqtt using the respective add-ons. It got really bad to a stage where nothing worked, mosquitto taking 99% cpu all the time. After a lot of struggling I managed to have 5.1.1 working with only zigbee2mqtt but I still see many errors accumulating just not enough to break the system. 

No ACK from MQTT server in 10 seconds (mid: 7)
2:04:42 AM – MQTT (WARNING) - message first occurred at 2:04:38 AM and shows up 11 times
Setup of switch platform mqtt is taking over 10 seconds.
2:04:33 AM – Switch (WARNING)

1614953178: Socket error on client <unknown>, disconnecting.
1614953242: New connection from 172.30.32.1 on port 1883.
1614953242: Socket error on client <unknown>, disconnecting.
1614953362: New connection from 172.30.32.1 on port 1883.
1614953362: Socket error on client <unknown>, disconnecting.
1614953482: New connection from 172.30.32.1 on port 1883.

As said the reconnecting goes on and eventually seems to kill the broker at it does not allow connections to work any longer.

**Comment by alex-savin** [comment]
As one person mentioned, enough is enough! Agreed!
Decided to stop using all the critical addons with "magic" from HA dev team. It is all the time gambling will it work after any update.
Setup a cluster of three VerneMQ instances with HAProxy as a load balancer.

**Comment by RichieFrame** [comment]
@michaelarnauts 
> One of the changes between 5.1 and 5.1.1 is this: [d291f56](https://github.com/home-assistant/addons/commit/d291f564ced1c19d0cfaa349b9f3b4b91e615151) So it is probably related to that, although I don't know what bashio does.

I did an analysis of the docker hub images for 5.1 and 5.1.1 and found some interesting things, posted the details in the HA forum thread about this

https://community.home-assistant.io/t/mosquitto-5-1-1-is-broken/286979/10

**Comment by michaelarnauts** [comment]
I'm afraid my issue is solved. I've removed the /data/mosquitto.db file, and it stopped saving message to memory. I assume there was once a client that had connected with the durable flag, and it subscribed to a heavy traffic topic. I think mosquito will keep all messages in case that client reconnects.

It might be worthwhile to check if this can fix this issue in your case...

**Comment by christoph-luebbe** [comment]
This does not solve the mentioned issues. I just checked this by deleting mosquitto.db in the container and then made the update to 5.1.1. MQTT broke instantly in HA. devices were not available.

**Comment by Eeeeeediot** [comment]
> For those who want to rollback 5.1:
> 
> * Backup and uninstall mosquitto 5.1.1
> * Fork mosquitto repository, edit "version": "5.1.1" to "version": "5.1." in config.json
> * Add this custom repository in the supervisor's add-on store and install
>   I installed mosquitto 5.1 newly with this method, hope this help

hey i tried this but it didnt allow me to add the fork as a  custom repo

**Comment by Mariusthvdb** [comment]
> I'm afraid my issue is solved. I've removed the /data/mosquitto.db file, and it stopped saving message to memory. I assume there was once a client that had connected with the durable flag, and it subscribed to a heavy traffic topic. I think mosquito will keep all messages in case that client reconnects.

glad you got it working, but this can't be a solution for those of us on Home Assistant OS unfortunately. Hope the update will get fixed anytime soon, did we see some response lately from the dev team? Just so we know it is under progress?

**Comment by nickrout** [comment]
> did we see some response lately from the dev team?

The thread isn't that long! 

How did removing mosquitto.db work for you?

**Comment by Mariusthvdb** [comment]
that's just the thing, Home Assistant OS users cant delete a mosquito.db, because it isn't even there ? that is to say, in the samba reachable folders

**Comment by nickrout** [comment]
I haven't checked via samba and probably can't until I am on my LAN at home.

**Comment by Mariusthvdb** [comment]
cool if you would, it is a most unpleasant state fo affairs we're in right now. Needing the update apparently (why else would it be available) , but not being able to...

**Comment by michalk-k** [comment]
Guys it's 2 weeks since the issue has been identified and confirmed that v5.1 is working properly why v5.1.1 hasn't been rolled back yet? 
I assume more and more people are being affected by this issue.

**Comment by realthk** [comment]
> I assume more and more people are being affected by this issue.
Probably still not enough people. I suspect this error occurs only under some conditions not in every setup, otherwise it'd be investigated by now.

**Comment by christoph-luebbe** [comment]
I think I have a path to overcome the problem. The comment from realthk got me thinking and I used the following procedure to get 5.1.1 working for me.

1. Take a snapshot (complete) from your current home assistant instance
2. go to the mosquitto add-on page and copy your configuration into an text editor for later use.
3. deinstall the mosquitto add-on 
4. install mosquitto 5.1.1 and copy the configuration back from the editor
5. start the add-on

This solved the problem in my installation

**Comment by Mariusthvdb** [comment]
that could well be, but, before even trying this, why not wait for the dev's to solve the issue? Or, put another way, why update in the first place at all, what do we gain by doing so, does anyone have an idea of the new features/functionality/advantages of 5.1.1 over 5.1.0 ?

**Comment by christoph-luebbe** [comment]
Agreed. But I suspect that following releases will carry on the problem unless you cleanup your installation

**Comment by quarcko** [comment]
For me - reinstalling the addon does not helped.
I mean like initially, yes, there are no warnings, but after restarting zigbee2mqtt and zwave addons
so they both re-add autodiscovery topics - you are back to same situation as before.

**Comment by christoph-luebbe** [comment]
thats interesting. I had the exact situation with zwaveJS2mqtt and it is now running for over 12h as with 5.1

**Comment by quarcko** [comment]
I must add that i still use old OpenZWave addon as i had no time to migrate yet.
So maybe that's the problem?
Did you used OpenZWave in the past?
Maybe for you it helped because basically reinstall removed OpenZWave topics that are problematic?
and for you they did not return, as you use now newer zwaveJS2mqtt
and for me problem is back because OpenZWave is still there...

just assumming...

**Comment by christoph-luebbe** [comment]
I used zwave2mqtt before and did the migration to zwaveJS2mqtt 4 weeks ago. So your assumption might be correct.

**Comment by chumbazoid** [comment]
FWIW, I've also experienced this with a Debian 10 x64 (supervisor supported) installation. Have never used Z-Wave / Zigbee and broker delete+reinstall did not resolve. Also tried deleting and reinstalling the MQTT integration.

With reinstalls, HA either could not connect to the broker at all (per HA log) or repeatedly connected+disconnected due to "socket error" (per broker log). Reversion to a 5.1 snapshot always restored normal functionality.

I'm abandoning the addon broker and followed [this guide](https://www.digitalocean.com/community/tutorials/how-to-install-and-secure-the-mosquitto-mqtt-messaging-broker-on-debian-10) to install MQTT directly on Debian. It was painless and quick with no functionality loss; only needed to rename some entities on my dashboard for cosmetic purposes. I suppose this setup is no longer supervisor compliant but the new broker seems snappier than ever.

**Comment by nickrout** [comment]
@chumbazoid I believe you are the only person on debian+supervised to report being affected by this. (Although most people unhelpfully don't say)

**Comment by alex-savin** [comment]
@nickrout he is not the one!

**Comment by nickrout** [comment]
What?

**Comment by realthk** [comment]
Probably the "only" is missing, as @chumbazoid is certainly not the _only_ one: I'm also using HA Supervised on Debian 10, but have no time to experiment with removing-reinstalling MQTT (and also use a few topics to store information with retained messages), I'm fine with 5.1 for now.

**Comment by jsb5151** [comment]
My system is also a supervised deployment.

**Comment by jysaloma** [comment]
Having this issue as well. High CPU usage

<img width="1211" alt="kuva" src="https://user-images.githubusercontent.com/24212289/112101331-20289400-8baf-11eb-8021-23338f47963c.png">

and logs gets this 
[08:08:44] INFO: [INFO] found tasmota3 on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
1616479725: Socket error on client <unknown>, disconnecting.
[08:08:59] INFO: [INFO] found tasmota3 on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
[08:09:09] INFO: [INFO] found tasmota3 on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
[08:09:24] INFO: [INFO] found tasmota3 on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
1616479765: Socket error on client <unknown>, disconnecting.
[08:09:35] INFO: [INFO] found zigbee2mqtt on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
[08:09:42] INFO: [INFO] found zigbee2mqtt on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
[08:09:56] INFO: [INFO] found zigbee2mqtt on local database
/bin/auth_srv.sh: line 17: echo: write error: Broken pipe
1616479805: Socket error on client <unknown>, disconnecting.

**Comment by jysaloma** [comment]
Re-install didn't help. I am having 18 local users in configfile 
logins:
... [18 local users]

anonymous: true
customize:
  active: true
  folder: mosquitto
certfile: fullchain.pem
keyfile: privkey.pem
require_certificate: false

access files set as

mosquitto $ cat acl.conf 
acl_file /share/mosquitto/accesscontrollist

and all users set in accesscontrollist
mosquitto $ cat accesscontrollist 
user zigbee2mqtt
topic readwrite #
...

**Comment by jysaloma** [comment]
Not sure if these are connected but I removed the Zigbee2mqttAssistant addon and rebooted my PI. Now situation seems to be normal and CPU load decent
<img width="1015" alt="kuva" src="https://user-images.githubusercontent.com/24212289/112129886-87a30b80-8bd0-11eb-8b14-6d879c3a7b24.png">

**Comment by cogneato** [comment]
> that's just the thing, Home Assistant OS users cant delete a mosquito.db, because it isn't even there ? that is to say, in the samba reachable folders

Uninistalling the addon and reinstalling would remove the mosquitto.db

**Comment by nickrout** [comment]
> Uninstalling the addon and reinstalling would remove the mosquitto.db

It should NOT be necessary to remove mosquitto.db. This is a bug. It needs to be fixed. Why is there no dev input on this?

**Comment by maddoxjny** [comment]
uninstalling and reinstalling does not fix the problem.   High cpu and devices not functioning.

**Comment by nickrout** [comment]
If you want to remove mosquitto.db (at your own risk) then there are two ways to do it that I can see.

**Log into the base operating system and remove the file**

```shell
rm /usr/share/hassio/addons/data/core_mosquitto/mosquitto.db
```

Or

**Enter the addon container and remove the file**
```shell
docker exec -it addon_core_mosquitto /bin/bash
rm /data/mosquitto.db
```

**Comment by Bonuscoupon** [comment]
> 
> 
> I think I have a path to overcome the problem. The comment from realthk got me thinking and I used the following procedure to get 5.1.1 working for me.
> 
>     1. Take a snapshot (complete) from your current home assistant instance
> 

In order to do this I re-installed it and re-started it - and then 5.1.1 worked fine. I don't know whats different to the situation before but it works now....

**Comment by christoph-luebbe** [comment]
Yes, I had the exact same experience. Also scratching my head why it works now.

**Comment by CubieMedia** [comment]
Thanks for the tips, i also reinstalled the addon but it did not help with the issue on my installation. Also deleted the database manually, still no luck!

I can reproduce the problem very easily with restarting the addon while connected with MQTT-Explorer and trying to send messages as soon as it reconnects. It takes quite some time until the messages get excepted.

Can anyone with the problem verify this?

**Comment by michalk-k** [comment]
do you mean v5.1 doesn't work for you?

**Comment by CubieMedia** [comment]
i tested with version 5.1.1, did i get this wrong?

were these tip all for version 5.1 ... i missed that in the thread.

**Comment by va3jme** [comment]
Sigh.... Rolled back to 5.1, and still can't get my Sonoff Zigbee bridge to work :(

**Comment by raidnet-ms** [comment]
Esphome not working while deconz things do. Curious...

**Comment by buzztiaan** [comment]
For reference ; https://community.home-assistant.io/t/mosquitto-5-1-1add-on-is-broken/286979/11

It seems 5.1.1 changed a LOT of stuff instead of just the mosquitto broker

**Comment by NetJaro** [comment]
> For those who want to rollback 5.1:
> 
> * Backup and uninstall mosquitto 5.1.1
> * Fork mosquitto repository, edit "version": "5.1.1" to "version": "5.1." in config.json
> * Add this custom repository in the supervisor's add-on store and install
>   I installed mosquitto 5.1 newly with this method, hope this help

Hi.

I try this but I cant add custom repo to supervisor

> 21-04-03 07:36:20 ERROR (MainThread) [supervisor.store.git] Can't clone https://github.com/NetJaro/addons/tree/master/mosquitto/ repository: Cmd('git') failed due to: exit code(128)

**Comment by Mariusthvdb** [comment]
can confirm that, upon cogneato's suggestion in Discord, deleting the 5.1 add-on, (copying the config) and re-installing (the now new 5.1.1) Add-on with the copied config, everything is running smoothly. NO errors in the log, and all topics are live.

So, don't update, but re-install which gets you the new version (and rewrites the mosquito.db) which seems to be the issue.

**Comment by bsmeding** [comment]
I also can confirm that removing and re-adding the add-on works well. 
Please note the configuration first. Remove add-on, wait a minute so local storage is removed.

I also restarted home assistent. Then installed add-on, check if username and password is in configuration of mosquitto and the mqtt integration (see configuration of mqtt in the integrations page, of configuration.yaml if configured via .yaml)

**Comment by jrhbcn** [comment]
Hello, 

@Mariusthvdb @bsmeding Unfortunately that is not my experience. I did remove my add-on and installed it from scratch. The MQTT disconnections I was suffering (with my [shelly](https://shelly.cloud/) devices) did go away compared with upgrading normally from 5.1 to 5.1.1. However, with add-on version 5.1.1, MQTT seems to be much more slow, connecting to the broker with [MQTTExplorer](http://mqtt-explorer.com/) takes significantly more time with 5.1.1 than 5.1. Also core logs get flooded with "No ACK from MQTT server in 10 seconds" errors as in [here](https://github.com/bieniu/ha-shellies-discovery/issues/116). I have not seen any other log entries (not in core of from add-on) between 5.1 and 5.1.1 that might point to some explanation but if any more logs are needed to help track this problem I am happy to help.

I am using Home Assistant OS on a raspberry pi 4. This is my config info:

## System Health
version | core-2021.4.3
-- | --
installation_type | Home Assistant OS
dev | false
hassio | true
docker | true
virtualenv | false
python_version | 3.8.7
os_name | Linux
os_version | 5.4.83-v8
arch | aarch64
timezone | Europe/Madrid

## Home Assistant Supervisor

host_os | Home Assistant OS 5.13
-- | --
update_channel | stable
supervisor_version | supervisor-2021.03.9
docker_version | 19.03.15
disk_total | 457.7 GB
disk_used | 23.7 GB
healthy | true
supported | true
board | rpi4-64
supervisor_api | ok
version_api | ok
installed_addons | Samba share (9.3.1), chrony (2.0.2), Home Assistant Google Drive Backup (0.103.1), Grafana (6.3.0), AdGuard Home (4.0.0), InfluxDB (4.0.4), ESPHome (1.16.2), TasmoAdmin (0.14.1), Zigbee2mqtt (1.18.1-1), motionEye (0.11.1), WireGuard (0.5.1), AppDaemon 4 (0.5.2), Visual Studio Code (3.3.0), Terminal & SSH (9.1.0), File editor (5.2.0), Check Home Assistant configuration (3.6.0), JupyterLab (0.5.0), Z-Wave JS (0.1.17), Mosquitto broker (5.1), Network UPS Tools (0.6.2)

</details>

**Comment by bsmeding** [comment]
Does the log from mqtt addon give more information?
Also does all the devices in mqtt topics be slower or only specific ones?

I have almost all devices connected via mqtt (zigbee, zwave, milight) after
you’re message i tried several in the homeassistant topic and cannot find
any slower responding devices.

Dis you reconfigure mqtt server in the mqtt integration?

Op zo 11 apr. 2021 om 19:01 schreef jrhbcn ***@***.***>

**Comment by jrhbcn** [comment]
> Does the log from mqtt addon give more information? Also does all the devices in mqtt topics be slower or only specific ones? I have almost all devices connected via mqtt (zigbee, zwave, milight) after you’re message i tried several in the homeassistant topic and cannot find any slower responding devices. 

Not that I can think of, both logs seem normal indicating connections from clients normally. I use MQTT with both shelly, zigbee and others (system_sensors) with more than 50 devices. It seems to be more related to shelly devices than zigbee or others (but haven't checked 100% so it is more a feeling). 

> Dis you reconfigure mqtt server in the mqtt integration?

Yes, actually one of my 5.1 <-> 5.1.1 installations (do not remember when) the default config stopped working so I had to create a new user in the "logins" section and re-configure the MQTT integration with that manual user. 

Thanks for looking into this!!

**Comment by Algirdyz** [comment]
I'm having similar issues. But I cannot compare to an older installation because I just started using Home Assistant this weekend.
I'm trying to connect arduino MQTT devices with the MQTT broker and initially they connect successfully with messages being sent in between.  

However after a few minutes it disconnects and Arduino cannot connect to it anymore until I reset it.
When trying to reconnect, Arduino is just waiting for a response, in the mean time MQTT broker logs say that the device connected successfully. After some time, depending on my timeout setting broker states the device exceeded timeout. At that point arduino also says that the server exceeded timeout and it just loops this way. So it seems to me that arduino expects some sort of response which doesn't come from the broker.

This is not an issue with the network because it works perfectly on first attempt after each reset and never succeeds when attempting to reconnect.

My arduino also did not disconnect when I tried setting up a public broker instead of using the adding. 

I've tried everything I could find online and right now this seems to be the only open end left, which i cannot test since I can't downgrade the addon.

**Comment by Mariusthvdb** [comment]
> Hello,
> 
> @Mariusthvdb @bsmeding 
 However, with add-on version 5.1.1, MQTT seems to be much more slow, connecting to the broker with [MQTTExplorer](http://mqtt-explorer.com/) takes significantly more time with 5.1.1 than 5.1. Also core logs get flooded with "No ACK from MQTT server in 10 seconds" errors as in [here](https://github.com/bieniu/ha-shellies-discovery/issues/116).

I dont see these errors in the logs anymore, but I can confirm the very long connection time to MQTT explorer. This was almost immediate before, and now takes 45 seconds. Have only 3 clients for the broker, 1 off which is my Owntracks which hasn't even been active yet. the other is a HA integration (on the same HA instance and the add-on broker) with some bluetooth trackers, the other is a dedicated Zwave hub.

Ill try to downgrade once more to see if this helps.

**Comment by grantalewis** [comment]
Thanks to those of you who have suggested methods for getting things working on 5.1.1. I've tried three times and so far haven't had any luck. I think I've tracked the problem down to the HA MQTT Integration not seeing the new broker but am stuck beyond that. Hoping someone might be able to give me a nudge.

Here are the steps I've taken so far:

Supervisor | Dashboard | Mosquitto broker | Configuration | (save info there to .txt file)
Supervisor | Dashboard | Mosquitto broker | Info | Uninstall
(reboot system)
Supervisor | Add-on Store | Search for mosquitto | reinstall
(reboot system)

I seem to see favorable info in Supervisor | Dashboard | Mosquitto broker | Log, but after waiting at least 30 minutes, I've got MANY entities that are unavailable. I can use MQTTExplorer to log on to the broker and see that my (cached?) entities are there, but they just show as "unavailable" in HA. 

Under 5.1 if I go to Configuration | Integrations | MQTT | Configure | Re-configure MQTT, leave the defaults, and click Submit, I'm able to set various options, click Submit, and see the "Success!" message.

But after upgrading to 5.1.1 if I go to Configuration | Integrations | MQTT | Configure | Re-configure MQTT, leave the defaults (which should be, I'm assuming, the same settings that work with 5.1), I see "Failed to connect."

So I think I'm seeing that the problem lies not with  Mosquitto broker version 5.1.1 but perhaps with the Integration located in Configuration.
  
Ideas? Suggestions? Thanks.

**Comment by johnjoemorgan** [comment]
@grantalewis here's a few things that come to mind 

You may have done this but on the 2 occasions that I've had a total breakdown of the Zigbee2mqtt (devices present but no MQTT messages reaching the HAS) I've found that re-flashing the coordinator works to 'reboot' the system. Maybe this is the 'nudge' your problem needs to be solved?  

Thinking about pushing other stuff at the Mosquito Broker and HAS:

Can you run any other Service2mqtt such as Z-Wave 2 MQTT - I'm guessing not otherwise you'd have said/  So moving on ... 

Perhaps installing a MQTT service such as 

https://gitlab.com/iotlink/iotlink

on a Windows PC and testing if you can scrape the MQTT messages from that to the HAS successfully.  That will help figure out id it's just Zigtbee2mqtt or all MQTT messages?

**Comment by jrhbcn** [comment]
> But after upgrading to 5.1.1 if I go to Configuration | Integrations | MQTT | Configure | Re-configure MQTT, leave the defaults (which should be, I'm assuming, the same settings that work with 5.1), I see "Failed to connect."

@grantalewis, Do you have a user defined in "logins" section of the configuration of the addon? and using that user in the HA mqtt integration. Something similar happened to me when testing 5.1 <-> 5.1.1 installation and at some point I had to manually set it up (I guess it was automatically done when I installed HA for the first time).

**Comment by bsmeding** [comment]
@grantalewis try to change the username and password in both the addon and the integration page

I found in the code that there was a default homeassistant user and password in mqtt addon and think that is removed or broken. Mu setup prior to 5.1.1 had a homeassistant user in the integration but that username and password was not visible in the addon configuration

Reset on both sides maybe solve the connection issue

**Comment by srnoth** [comment]
After reverting to 5.1 to resolve issues, decided to give 5.1.1 another try based on some of the suggestions here, and so far so good. Here's what I did:
-Backup backup backup (lol)
-Stopped zigbee2mqtt plugin and set it not to start on boot
-Uninstalled mosquitto plugin
-Deleted HASS local user which I was using for MQTT (mosquitto) auth
-Rebooted the entire HASSIO box
-Installed mosquitto plugin (5.1.1), adjusted configuration to use a login specified there instead of a HASS user. 
-Started mosquitto plugin
-Started monitoring "#" on "listen to topic" so I could observe all messages coming in
-Restarted all Tasmota devices, waited for each of them to check in and finish re-adding retain messages
-Started Zigbee2mqtt plugin, set to start on boot, observed messages start coming in, waited until all retain messages were re-added (hundreds in my case, took few minutes)
-Once the flood of initial messages stopped, I tested my devices and observed none of the slowdown or latency that happened when I did the first in-place 5.1 --> 5.1.1 update.
-Only seeing a socket error from one device which I will need to investigate separately, but no other errors so far.

**Comment by jrhbcn** [comment]
@srnoth  have you tried connecting to the mqtt server using [MQTT-Explorer](http://mqtt-explorer.com/)? For me using v5.1 works OK but on version v5.1.1 just takes ages to connect and start showing messages (6-7 seconds vs 45 seconds). Furthermore, all my zigbee2mqtt also seem to work ok with v5.1.1 but not my shelly devices (through mqtt [shelly discovery script](https://github.com/bieniu/ha-shellies-discovery)) than report ACK timeouts in the logs.

**Comment by srnoth** [comment]
> @srnoth have you tried connecting to the mqtt server using [MQTT-Explorer](http://mqtt-explorer.com/)? For me using v5.1 works OK but on version v5.1.1 just takes ages to connect and start showing messages (6-7 seconds vs 45 seconds). Furthermore, all my zigbee2mqtt also seem to work ok with v5.1.1 but not my shelly devices (through mqtt [shelly discovery script](https://github.com/bieniu/ha-shellies-discovery)) than report ACK timeouts in the logs.

Just tested and it connects and starts displaying messages within a couple seconds. Takes a but longer than that to load all.

**Comment by grantalewis** [comment]
@johnjoemorgan I gave Z-Wave 2 MQTT a couple of test runs a few weeks back but my entities were missing attributes. I decided waiting was wiser.

@jrhbcn I did have some defined, yes, and as you and @bsmeding pointed out, making sure that the credentials matched on both sides did allow the broker to connect to the add-on. Score 1 for you both. BUT -- after successfully performing the upgrade and waiting for an hour, clicks were taking 5+ MINUTES to finally register, no exaggeration. I was seeing literally tens of thousands of messages in MQTTExplorer. The system was completely choked, unusable. Rolling back to 5.1, everything goes back smooth as butter. It's crazy.

**Comment by thecode** [comment]
I know it is a long shot, but do you have authentication enabled for the MQTT broker? I had problems with brokers with authentication disabled, I have no clue why, but connection took very long time.
 BTW, I am tracking this issue since on 5.1.1 I have higher CPU usage, but that's all, no delays, MQTT explorer connects within 2 seconds. My broker was upgraded from 5.1 to 5.1.1 when it was released, was never tried to reinstall it.

**Comment by jrhbcn** [comment]
@grantalewis @thecode what hardware are you running HA? Could it be that the high CPU causes slow downs on limited systems (such as raspberry 4) but it does not show on other more powerful systems? Didn't check the CPU utilisation myself on 5.1.1

**Comment by grantalewis** [comment]
Don't think it's a hardware thing. I'm running Ubuntu, HA Supervised (Docker) on a Mac mini Core 2 Duo 2.66.

**Comment by thecode** [comment]
> @grantalewis @thecode what hardware are you running HA? Could it be that the high CPU causes slow downs on limited systems (such as raspberry 4) but it does not show on other more powerful systems? Didn't check the CPU utilisation myself on 5.1.1

I am running on VM using 4 cores of AMD 3700X CPU, one of the problems using such setup is indeed that I can't compare my system to other systems.  When a RPI3 will get clogged by high CPU usage, I hardly notice this on my system :-).

**Comment by srnoth** [comment]
> I know it is a long shot, but do you have authentication enabled for the MQTT broker? I had problems with brokers with authentication disabled, I have no clue why, but connection took very long time.
> BTW, I am tracking this issue since on 5.1.1 I have higher CPU usage, but that's all, no delays, MQTT explorer connects within 2 seconds. My broker was upgraded from 5.1 to 5.1.1 when it was released, was never tried to reinstall it.

I am noticing slightly higher latency with this new version as well. A typical work flow of mine is turning on a zigbee light from a zigbee remote. What is weird is if I haven't turned a light on/off for some time, when I first press the on or off button on the remote, there is a noticeable (~1s) delay before the light responds. But, having just used the remote, if I then press a button a few seconds later, the response is almost instantaneous (as was always the case with 5.1).

Not sure if there is something with the new version that causes the MQTT server to go to sleep, resulting in that first action being slower. Or, if there is a different timeout on the connection between zigbee2mqtt and the broker, resulting in zigbee2mqtt having to reconnect when the first action is submitted in some time.

**Comment by nickrout** [comment]
I have to say that an initial connection being slow reminds me of reverse
DNS issues.

Client connects to server
Server does reverse DNS lookup
Reverse times out ****
Connection continued

The delay occurs here ****

If course I am just guessing, but this is a common server connection
problem.

**Comment by srnoth** [comment]
Just did a direct comparison to 5.1. While 5.1.1 is usable under clean install, it still is a noticeable regression from 5.1.

E.g. listening to the "#" topic on 5.1, clicking "Start Listening" almost immediately starts processing hundreds of messages. Same action in 5.1.1 takes several seconds to even respond, and then appears to go through messages one at a time, taking several minutes to process through 1000 messages.

**Comment by nickrout** [comment]
Hardly a scientific test. Try a test using identical messages and something
command line like mosquitto_sub that you can time accurately.

**Comment by srnoth** [comment]
@nickrout unfortunately with individual messages, the delay is not enough to measure. We're talking about less than a second in latency (but still enough to notice the difference between pressing a button, and lights turning on immediately, vs. a slight delay).

But when compared to loading all messages (I usually have over a thousand at any given time) the delay is noticeable. 5+ seconds vs. almost instant.

For those of us who are not developers, we're just trying to provide enough info to show that _**something**_ changed between 5.1 and 5.1.1 that has made things incrementally slower. Not enough to notice with a fresh database and just a few devices, but start adding an entire house of smart devices and the delay adds up.

**Comment by jeunesto** [comment]
> > For those who want to rollback 5.1:
> > 
> > * Backup and uninstall mosquitto 5.1.1
> > * Fork mosquitto repository, edit "version": "5.1.1" to "version": "5.1." in config.json
> > * Add this custom repository in the supervisor's add-on store and install
> >   I installed mosquitto 5.1 newly with this method, hope this help
> 
> hey i tried this but it didnt allow me to add the fork as a custom repo

I'm really struggling for days with this issue since upgrading to 5.1.1, I've all my z-wave devices that are unavailable. I think I've tried every solutions mentionned here to get it working without success. Unfortunately, no back-up old enough to get me back with the 5.1 version. I've tried to fork the repository but then add it does not work at all. I'm stuck. Countless hours lost to solve this issue... Too many automation in the house so this is huge problem. As it's been week that the problem is around, I do suppose it will not be solved by an upgrade. If someone can point me out how to get back to 5.1, that would be very great. Thanks in advance.

**Comment by CubieMedia** [comment]
There are several options mentioned in this thread. 

'docker pull homeassistant/aarch64-addon-mosquitto:5.1'

@hung2kgithub
has a solution with an own repository somewhere i the middle of this thread

please just check the whole thread if you need a solution for you own situation

**Comment by balilu91** [comment]
Just putting my experience here with this issue.
I started having the sluggish experience same as everyone else is reporting. I'm running on a PI4. 

Following some of the above comments I removed the addon to clear the musquitto.db and re installed it and it worked fine for about two/three weeks.

Then the issue started cropping up again slowly and today it got to a point I had remove the addon once again.

Before removing the addon I had a look using MQTT Explorer and noticed $sys/broker/messages/stored had over 900k

Never looked at it before but it's seems a bit too high is it expected to maintain so many message?

**Comment by grantalewis** [comment]
[I experienced the same](https://github.com/home-assistant/addons/issues/1887#issuecomment-818855421), @balilu91. Something is definitely screwy.

**Comment by niemyjski** [comment]
I've been running into the same behaviors described here.. I just removed my addon and readded with the same configuration. Now nothing works:

```
[15:14:33] WARNING: [WARN] Not found homeassistant on local database
1618604073: Socket error on client <unknown>, disconnecting.
1618604193: New connection from 172.30.32.1 on port 1883.
1618604193: Socket error on client <unknown>, disconnecting.
1618604313: New connection from 172.30.32.1 on port 1883.
1618604313: Socket error on client <unknown>, disconnecting.
1618604433: New connection from 172.30.32.1 on port 1883.
```

**Comment by fabricepipart** [comment]
I also had to revert to 5.1 to have a functional setup (otherwise I had crazy delays each time I was switching on a light). This seems like an important issue. And I am afraid, there is way more than you might think between the 5.1.1 update.
Unfortunately, I don't see this issue getting a lot of attention. Is there anything we can do to clarify or reproduce the issue? How can we help getting this solved?

**Comment by DeviousVon** [comment]
Unfortunately I don't have the ability to roll back as all my snapshots were partial (that won't happen again).  I'm unable to add any new devices (Tasmota) even though it looks good in the console.  Frustrating.  Really not sure what to do about it.  I've considered installing mosquitto on a different machine, but I'm using HA authentication. I've not been able to fix it with the uninstall reinstall thing others are saying works, it just goes back to not working. I'm willing to work with anyone that can address it, I'm unfortunately not a developer...  At this point I'm considering downloading an old version and starting over (this is a LAST resort as I've put a lot of time into HA).  It sucks having to stand up and walk ALL the way across the room to turn lights on and off.

**Comment by grantalewis** [comment]
@DeviousVon You say "download an old version and starting over." I'm assuming you mean Home Assistant? Note that the Mosquitto broker, the component with the problem we're discussing here, is an add-on to Home Assistant, not Home Assistant itself. Even if you can download an old version of HA, you'd still need to be able to download and install an older version of the add-on (5.1 or earlier).

Different topic, "starting over." If you do somehow manage to fix the problem by "downloading an old version," it's not necessary to start over. Just make sure you back up everything first. You'll need access to Home Assistant's filesystem -- I use SMB.

1) First do a complete snapshot of your current setup. The .tar file will be saved in Home Assistant's backup folder. Save that file to an external location.
2) Then for safety I'd recommend also doing a manual backup. Start by making copies of any/all .yaml files and saving those files to an external location, too. Depending on how your system is configured there could be many, but they'll all be small in size. I have upwards of 50 individual .yaml files, but that's because I've split my config up.
3) If you have any Z-Wave xml files on your system (zw*.xml), back those up, too.
4) If you have any Node-RED automations, manually back those up:
   - Open Node-RED
   - In the upper right, click the hamburger menu in the black border
   - Click Export | All flows | Copy to clipboard
   - Paste the clipboard contents to a text file and save to an external location
4) Manually back up your lovelace config:
   - From any page in Home Assistant click the three vertical dots in the upper right and click Edit Dashboard
   - Click the same three dot menu again and then click Raw configuration editor
   - CTRL-A to copy everything to to the clipboard
   - Paste the clipboard contents to a text file and save to an external location

Sounds like a lot of trouble, but it's really not. Takes 5-10 minutes.

So you'll have the .tar file as well as individual .yaml (etc.) files once you're done. After re-installing Home Assistant, first try copying the large .tar file to the same backup folder where you found it and doing a restore in Home Assistant: Supervisor | Snapshots. If anything fails or you otherwise have problems, you'll have the manually-backed-up files to turn to. Just more or less reverse the steps you used to back up everything.

It's not necessary to start over. You can easily go from brand-new install to fully restored config in well under an hour.

**Comment by DeviousVon** [comment]
Thank you for the eloquent response!  I did end up getting the 5.1 installed (forked the addons and downgraded the version) Unfortunately it didn't fix the issue I was experiencing.    So I've taken your advice!  Backup and reinstall.  Thank you

**Comment by grantalewis** [comment]
You're welcome. When you do a Snapshot restore in HA, you have the option to disable portions of what's to be restored from the Snapshot. This is how a person could revert to 5.1. After installing the buggy 5.1.1, make a backup of everything (for safety), and then use an earlier Snapshot to restore only the 5.1 version Mosquitto broker. That process won't work if you don't already have a Snapshot with 5.1 on it, of course, but it's good to know it's there for the next time.

**Comment by geoffrey-vl** [comment]
I'v' also had issues with this MQTT addon on the brand new install of the HassOS on a RaspberryPI 3.

For now I'm running the deprecated HiveMQ addon, it's also working well.

**Comment by Murder3D** [comment]
Me too.

HA on SSD on RPi4 4GB.

Here's the thread on HA Community --> https://community.home-assistant.io/t/tasmotized-devices-slow-response/302464/7

**Comment by fabricepipart** [comment]
It could help to narrow down where the issue comes from. I migrated my setup (tasmotized sonoff triggered by zigbee2mqtt xiami switches) from a Raspberry Pi 3 (where the fallback to 5.1 had worked) to a VM on my Mac. And on the x86 Mosquitto 5.1.1 works without any delay.

**Comment by frenck** [comment]
Reserved my Thursday for looking into this add-on.

**Comment by jacoscar** [comment]
I have noticed MQTT being slow on both a Raspberry Pi3B+ and a VirtualBox virtual machine (same snapshot)
Something must be wrong with MQTT as between a flick of a switch in the dashboard and the actual toggle of the lights there are sometimes 2 to 5 seconds, but then, after the first time, it is instantaneous
I don't know if other errors in my log are relevant, but I'll leave them anyway

`2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 47)
2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 48)
2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 49)
2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 50)
2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 51)
2021-05-04 09:32:15 WARNING (MainThread) [homeassistant.components.mqtt] No ACK from MQTT server in 10 seconds (mid: 52)
2021-05-04 09:32:16 ERROR (MainThread) [aiohttp.server] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 422, in _handle_request
    resp = await self._request_handler(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_app.py", line 499, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 56, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 18, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 74, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 135, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 131, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/http.py", line 43, in get
    return await WebSocketHandler(request.app["hass"], request).async_handle()
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/http.py", line 142, in async_handle
    await wsock.prepare(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_ws.py", line 135, in prepare
    payload_writer = await super().prepare(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_response.py", line 378, in prepare
    return await self._start(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_response.py", line 386, in _start
    await self._write_headers()
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_response.py", line 458, in _write_headers
    await writer.write_headers(status_line, self._headers)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/http_writer.py", line 119, in write_headers
    self._write(buf)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/http_writer.py", line 67, in _write
    raise ConnectionResetError("Cannot write to closing transport")
ConnectionResetError: Cannot write to closing transport
2021-05-04 09:35:54 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 09:35:54 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 09:40:47 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 09:40:48 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 09:50:51 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 09:50:51 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 10:29:39 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 10:29:40 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 10:55:18 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 10:55:18 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 11:12:32 WARNING (MainThread) [homeassistant.components.automation.new_code] New Code: Already running
2021-05-04 11:32:46 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 11:32:46 WARNING (MainThread) [homeassistant.components.mqtt.mixins] JSON result was not a dictionary
2021-05-04 11:47:10 WARNING (MainThread) [homeassistant.components.http.ban] Login attempt or request with invalid authentication from vodafone.broadband (192.168.0.1). (Mozilla/5.0 (iPhone; CPU iPhone OS 14_5 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Home Assistant/2021.4.1 (io.robbie.HomeAssistant; build:2021.115; iOS 14.5.0) Mobile/HomeAssistant, like Safari)`

**Comment by Klexx** [comment]
> [...] between a flick of a switch in the dashboard and the actual toggle of the lights there are sometimes 2 to 5 seconds, but then, after the first time, it is instantaneous.

I can confirm this behavior.

(running: core-2021.4.6, Mosquitto broker 5.1.1)

**Comment by frenck** [comment]
Made quite a bit of progress today, but I'm far from happy with the result at this point.
Will continue my journey in the next couple of days.

**Comment by seaniedan** [comment]
Thank you for looking, Frenck. I made some notes on 5.1.1 over at #1817 (scroll down!) - I hope it helps.

**Comment by seaniedan** [comment]
let me know if you need anyone to test!

**Comment by seaniedan** [comment]
If it helps, the issues I describe in #1817 were present in the 5.1 version of Mosquitto: I upgraded in an attempt to fix.

**Comment by jacoscar** [comment]
Is it possible to use an alternative mqtt broker in home assistant while this is being fixed?

**Comment by geoffrey-vl** [comment]
See https://github.com/home-assistant/addons/issues/1887#issuecomment-830625425

**Comment by dgreebe** [comment]
> Is it possible to use an alternative mqtt broker in home assistant while this is being fixed?

I just installed  the eclipse-mosquitto in docker, exposed the nessecary ports and its working perfectly for now. hope the add-on will be fixed asap..

for now, everything is up and running and working smoothly..

**Comment by matejzero** [comment]
> I just installed the eclipse-mosquitto in docker, exposed the nessecary ports and its working perfectly for now. hope the add-on will be fixed asap..

Did the same and it's now working really fast.

In the core addon, when I initially connected with MQTT Explorer, it needed 1-2s to connect and then sometimes 5s+ to get topics. Reconnect was faster.

This most likely also resolves my issues with automations not working: https://community.home-assistant.io/t/on-motion-wait-until-illuminance-sensor-is-updated

**Comment by frenck** [comment]
Status update time 🕐 

I've created quite a few versions of the add-on the past week and ended up with the simplest of them all 😓 


- MQTT :white_check_mark: 
- MQTT over WebSocket :white_check_mark: 
- MQTT with SSL :white_check_mark: 
- MQTT over WebSocket with SSL :white_check_mark: 
- Local users auth :white_check_mark: 
- Home Assistant users auth :white_check_mark: 
- Home Assistant integration discovery :white_check_mark: 
- Supervisor add-on services :white_check_mark: 
- Migrated everything to S6 overlay :white_check_mark: 
- Migrated everything to bashio
- Running on latest Alpine Linux :white_check_mark:
- Can handle a metric ton of auth requests in milliseconds per request :white_check_mark:

I'm happy with the result, it seems to perform better. Currently got Z-Wave JS 2 MQTT connected and around 25 ESPChip on ESPHome sending all a lot of status updates each second.

I'll leave that running overnight to see how it holds and do a final round of reviewing my own code in the morning.

**Comment by matejzero** [comment]
Do you maybe have numbers from the current mqtt addon to compare, to see if the poor results can be replicated in the old addon:)

**Comment by CubieMedia** [comment]
> Can handle a metric ton of auth requests in a couple of milliseconds per request

Very well done, this is hard to test as the problem always was only the first login (of each device) that took very long (> 5 sec).

Thanks for putting you head into this!

**Comment by matejzero** [comment]
@frenck if you connect to the MQTT server, do you get all topics right away? In the problematic server, my connection took some time (between 0-2s), but then I had to wait 5s-6s for topics to apear. Is that mitigated in your new version?

**Comment by frenck** [comment]
> if you connect to the MQTT server, do you get all topics right away?

Yes, however, this of course is affected by the amount of data you store, and how fast both the client and server are.

> Very well done, this is hard to test as the problem always was only the first login (of each device) that took very long (> 5 sec).


True! There is a big difference I. The handling now though, but will run a test without caching layers to measure the first-time every time.

**Comment by frenck** [comment]
@matejzero Auth/ACL caching disabled:

![image](https://user-images.githubusercontent.com/195327/118620290-a6cfab00-b7c5-11eb-9369-d35a8ea124b9.png)

**Comment by matejzero** [comment]
Doesn't seem to do much difference, around 3ms for local users and a few more when asking HA. This is fast!

>> if you connect to the MQTT server, do you get all topics right away?

> Yes, however, this of course is affected by the amount of data you store, and how fast both the client and server are.

I'm currently running the obsolete MQTT addon and compared to the core addon, it connects and shows all topics instantly compared to core addon where it takes a few seconds to get them. I wanted to do some mosquito_pub tests today but seeing as you got pretty far with the updated version, I'll just wait for the new release and test with that.

**Comment by CubieMedia** [comment]
With the HA user it looks like 8-10 ms and Auth/ACL disabled should create an environment like a restart and devices newly connect to the broker.

Test approved!

The 10ms are putting my hopes up that this will work soon again! Can i get any info of what you changed when the new version is released? Startup parameters from MQTT Broker or something container specific?

**Comment by frenck** [comment]
> Can i get any info of what you changed when the new version is released?

There is a changelog and, of course, it is open source: Yes, you can see that.

The main difference in this case, is that part of the authentication handling is now done by the NGINX webserver. That thing is able to handle a lot of requests: fast. This replaces the bash scripts piped to socat in the current version.

**Comment by CubieMedia** [comment]
> The main difference in this case, is that part of the authentication handling is now done by the NGINX webserver. That thing is able to handle a lot of requests: fast. This replaces the bash scripts piped to socat in the current version.

@frenck Thank you for that short inside.

**Comment by matejzero** [comment]
Will report results once the PR is merged and new version is available. Or is there an easy way to test it before it's merged?

**Comment by matejzero** [comment]
I did some quick testing and initial filing is that it works as it should now. MQTT client connects really fast and gets all it's topic at the same time.

Motion and button responses are also way faster then before, the same as with deprecated mqtt client from community addons.

**Comment by srnoth** [comment]
After a couple hours on 6.0.1, zero slow down issues.

**Comment by CubieMedia** [comment]
Yes it is working again, well done!

**Comment by Geoff571** [comment]
I'm happy to report things are looking a lot more responsive now that I've upgraded to 6.0.1, the errors in the logs and laggy switch/light responses seem to be gone. 

Many thanks @frenck

**Comment by bsmeding** [comment]
Yes working perfectly and fast! Thanks @frenck !

**Comment by pbzdyl** [comment]
Thank you @frenck - great job!

**Comment by jeffgraham-hub** [comment]
Is anyone else seeing errors in the supervisor log after upgrading to 6.0.1.? If I disable the watchdog on Mosquitto the errors quit. I also rolled back to 5.1.1 and verified no errors with watchdog enabled. 

```
21-05-21 04:59:12 ERROR (MainThread) [asyncio] Task exception was never retrieved
future: <Task finished name='Task-207448' coro=<Scheduler._run_task.<locals>._wrap_task() done, defined at /usr/src/supervisor/supervisor/misc/scheduler.py:58> exception=TypeError('an integer is required (got type str)')>
Traceback (most recent call last):
  File "/usr/src/supervisor/supervisor/misc/scheduler.py", line 62, in _wrap_task
    await task.coro_callback()
  File "/usr/src/supervisor/supervisor/misc/tasks.py", line 416, in _watchdog_addon_application
    if addon.in_progress or await addon.watchdog_application():
  File "/usr/src/supervisor/supervisor/addons/addon.py", line 479, in watchdog_application
    return await self.sys_run_in_executor(check_port, self.ip_address, port)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/supervisor/supervisor/utils/__init__.py", line 43, in check_port
    result = sock.connect_ex((str(address), port))
TypeError: an integer is required (got type str)
```

**Comment by jrhbcn** [comment]
@jeffgraham-hub I see the same error in the logs.

**Comment by realthk** [comment]
Same here, though still having errors like these after reverting back to Mosquitto 5.1, but only during startup, so probably from another addon (with MQTT 6.x the supervisor log was flooded with these even during the day, not only at startup)

**Comment by bsmeding** [comment]
Yes same errors

**Comment by Mariusthvdb** [comment]
these errors are gone/fixed in supervisor-2021.05.1

**Comment by frenck** [comment]
Please stop digging into old issues reporting unrelated new issues. Instead, just create a new issue for each individual thing so it can be tracked and looked at.

Thanks.

**Comment by jacoscar** [comment]
I upgraded to 6.0.1 and it was fine for a day but now I’m experiencing delays again

5.1 was fine though

**Comment by tylerszabo** [comment]
@frenck I've come up with a slightly less abrasive option for a similar comment. It's a bit more verbose but feel free to copy and paste it. The goal is to avoid discouraging good faith participation while still adhering to your preferred workflow. While it's not as efficient to add a buffer for the emotional state of the reader it can be considered a "social optimization" to add a bit more explanation.

"If this isn't resolved or there's a new problem please create a new issue for each individual thing so it can be tracked and looked at.

We have a workflow that involves opening a new issue instead of reviving old ones; this greatly helps with our tracking and organizing. Re-opening an issue or combining multiple subjects into a single issue adds complexity to the tracking."
